Filtering CLI

Dhruv-J opened the pull request:
This commit adds CLI functionality for filtering.

Signed-off-by: Dhruv Jain <2dhruv@gmail.com>
Co-authored-by: Stan Wong <swong394@gmail.com>

stanleywbwong commented on cmd/cli/main.go:
```diff
@@ -153,6 +193,14 @@ func runRollback(cmd *cobra.Command, args []string) {
+	getCmd.Flags().StringVarP(&until, "until", "u", "", "end of time range")
+	getCmd.Flags().StringVarP(&resource, "resource", "r", "", "resource name to filter by")
+	getCmd.Flags().StringVarP(&namespace, "namespace", "p", "", "namespace to filter by")
+	getCmd.Flags().StringVarP(&name, "name", "n", "", "name to filter by")
```
minor thing but maybe we should make namespace the -n flag to keep consistency with kubectl

stanleywbwong commented on cmd/cli/main.go:
```diff
@@ -81,6 +104,23 @@ var rollbackCmd = &cobra.Command{
+	}
+	return url
+}
+
```
does the simpler version of getURL I rewrote earlier work?

`func getURL() string {
	flags := []string{author, since, until, resource, namespace, name}
	flagnames := []string{"author=", "since=", "until=", "resource=", "namespace=", "name="}
	var parts []string
	for i, flag := range flags {
		if strings.TrimSpace(flag) != "" {
			parts = append(parts, flagnames[i]+flag)
		}
	}
	url := "http://localhost:" + port + "/changes?"
	url += strings.Join(parts, "&")
	return url
}`

stanleywbwong replied:
Does the simpler version of getURL I wrote before work?
```
func getURL() string {
	flags := []string{author, since, until, resource, namespace, name}
	flagnames := []string{"author=", "since=", "until=", "resource=", "namespace=", "name="}
	var parts []string
	for i, flag := range flags {
		if strings.TrimSpace(flag) != "" {
			parts = append(parts, flagnames[i]+flag)
		}
	}
	url := "http://localhost:" + port + "/changes?"
	url += strings.Join(parts, "&")
	return url
}
```

stanleywbwong commented:
We should probably add the server-side handler to this PR too

stanleywbwong commented on cmd/cli/main.go:
```diff
@@ -43,6 +46,26 @@ var rootCmd = &cobra.Command{
+var getCmd = &cobra.Command{
+	Use:   "get",
+	Short: "get changes by author, time range, and filepath",
+	Run: func(cmd *cobra.Command, args []string) {
```
also would be nice to add an example through in the `Example: ` field, see https://github.com/antrea-io/resource-auditing/pull/17/files

antoninbas commented on cmd/cli/main.go:
```diff
@@ -18,15 +18,20 @@ import (
 )
 
+// filter flags
+var getAuthor, since, until, resource, namespace, name string
```
why is it `getAuthor`? if there is a name conflict, you probably want to change the other definition to have some naming consistency here

Dhruv-J replied:
There wasn't a naming conflict, since there's a tagAuthor, I thought it would be better to specify which author it referred to.

antoninbas replied:
I see.. that's fine then. Maybe you could prefix all of the flags with `get` for consistency with the tag command?

antoninbas commented on cmd/cli/main.go:
```diff
@@ -91,6 +106,37 @@ var rollbackCmd = &cobra.Command{
+	flags := []string{getAuthor, since, until, resource, namespace, name}
+	flagnames := []string{"author=", "since=", "until=", "resource=", "namespace=", "name="}
+	var parts []string
+	for f, flag := range flags {
```
s/f/idx

Dhruv-J commented on pkg/webhook/webhook.go:
```diff
@@ -20,13 +20,19 @@ import (
+type Change struct {
+	Sha     string `json:"sha"`
+	Author  string `json:"author"`
+	Message string `json:"Message"`
```
They're all supposed to be capitalized, I'll fix this.

antoninbas replied:
actually, they are all supposed to *not* be capitalized (by convention I think)

antoninbas replied:
not addressed yet?

antoninbas commented on cmd/cli/main.go:
```diff
@@ -43,6 +48,16 @@ var rootCmd = &cobra.Command{
+	Run:   runGet,
+	Example: ` Getting changes by author and filepath
+    $ auditctl get -a kubernetes-admin -r k8s-policies -n default -f allow-client1.yaml
+    [{"sha":"a75dc67fd950b5ed052897b981c8d7b2cb05e9a5","author":"kubernetes-admin","Message":"Deleted K8s network policy default/allow-client1"}]
```
```suggestion
    [{"sha":"a75dc67fd950b5ed052897b981c8d7b2cb05e9a5","author":"kubernetes-admin","message":"Deleted K8s network policy default/allow-client1"}]
```

Dhruv-J replied:
This is just the way that the commit object is returned, which is what I print.

antoninbas replied:
this should no longer be the case after fixing the `Change` struct:

```
type Change struct {
	Sha     string `json:"sha"`
	Author  string `json:"author"`
	Message string `json:"message"`
}
```

The `json` tag determines the field name here. You may want to try again.

antoninbas commented on cmd/cli/main.go:
```diff
@@ -91,6 +106,35 @@ var rollbackCmd = &cobra.Command{
+	for idx, flag := range flags {
+		params.Set(flagnames[idx], flag)
+	}
+	url := "http://" + serverAddr + "/changes"
+	reqUrl := fmt.Sprintf("%s?%s", url, params.Encode())
```
one liner: `reqURL := fmt.Sprintf("http://%s/changes?%s", serverAddr, params.Encode())`

BTW, we avoid `reqUrl` as a variable name by convention, and we prefer `reqURL`. See https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Dhruv-J committed:
Filtering CLI

This commit adds CLI functionality for filtering as well as
the appropriate handlers.

Signed-off-by: Dhruv Jain <2dhruv@gmail.com>
Co-authored-by: Stan Wong <swong394@gmail.com>
